### ISSUE TITLE
Questions/suggestions for logging module

### ISSUE THREAD

**Issue by dkirkby** [desihub/desiutil#66]
I have some questions/suggestions about `log.get_logger()` after adapting `desisurvey` and `surveysim` to use it.

Environment variables usually set a default that can be overridden via the API, but `$DESI_LOGLEVEL` overrides the API `level` parameter.  I found this confusing and wonder if we should change this.

A command-line script should be able to set the log-level (e.g., via cmd-line options) and then have all library code respect this level.  I believe the way this is supposed to work is:
```
# script.py
log = get_logger(WARNING)

# module.py
log = get_logger()
log.info('...')
```
I found two issues with how this works in practice:
1. Calling `get_logger()` always sets the log-level to `INFO`, even after a previous call where a different level was specified.
2. Nothing prevents library code from calling `get_logger(DEBUG)` and changing the log level specified by the top-level script.  Library code should probably always call `get_logger()` but we could enforce this by raising an exception on attempts to change an already specified level. Even better, I think, would be to provide separate `get_logger()` (with no level arg) and `set_log_level(level)` methods.

Finally, it could be useful to support a hierarchy of DESI loggers with different levels.  For example, I might want to enable DEBUG messages from the arc calibration only, without being flooded with other DEBUG messages.  In practice, this could be implemented by passing an optional component arg to `get_logger()` that uses dot notation to express hierarchy.  The level of a child component would then default to the lowest level of any parent, and setting a parent level enforces a maximum level for all children.

**Comment by weaverba137** [comment]
Issue 1 is already fixed in PR #64, which will be merged relatively soon.

However, given issue 2 and the other comments, you're really just describing the built-in `logging` module itself.  In other words, I think the original intention of `get_logger()` was to *hide* all the options available in the `logging` module so that people who didn't know anything about it would still be able to write log messages.

I think this potentially opens up a debate larger than this ticket itself.  @sbailey, @tskisner, please chime in.  Perhaps we should briefly discuss on a telecon.

**Comment by tskisner** [comment]
I have no strong opinions on this.  In my experience with the pipeline, when running a large scale production one often wants lots of debugging info from all steps or minimal debugging for all steps.  If one is debugging a specific task (i.e. one particular frame that failed calibration), then obviously it is easy to set DESI_LOGLEVEL when interactively debugging that.  I have not used python logging much beyond the simple use case of setting the level for a whole process.

**Comment by weaverba137** [comment]
@sbailey, you said you wanted to comment on this issue, but I also really need to merge #64, & that affects this issue, at least in part:

> Calling `get_logger()` always sets the log-level to `INFO`, even after a previous call where a different level was specified.

That will no longer be the case after #64 is merged.

**Comment by sbailey** [comment]
Go ahead and merge #64.  I don't think that either hinders or helps the thoughts I'm trying to put together for logging.

**Comment by weaverba137** [comment]
OK, done, thanx!

**Comment by weaverba137** [comment]
Ping.  @sbailey, are you still composing your thoughts on this?

**Comment by sbailey** [comment]
I suggest the following, described in a user-facing way:
  1. minimal code to just get a standard format logger at the current default loglevel and use it
  2. ability to get a standard format but custom loglevel logger, e.g. to be used within a single function
     while not affecting the global state of other log usage
  3. ability to change the global default loglevel, e.g. for a script --debug option

And the same as (2) and (3) but for the option of including timestamps or not.

This stops short of the hierarchical logging levels that @dkirkby proposed which smell fancier than we need.  If someone wants to implement that and it doesn't impact the simplicity of 1-3, then ok.

I want (1) because I dislike the boilerplate of having to import `get_logger` just so that I can call it to get the object that I really need, but (2) and (3) are more important, i.e. the ability to separately control the local vs. global loglevel.

## In more detail ##

**Minimal overhead to just get a DESI standard format logger and use it**
i.e. even bypassing having to import `get_logger` just so that you can call it to get the logger that you really want to use:
```python
from desiutil.log import log
log.info('blat foo')
log.debug('biz bat')
```
This implies a pre-created singleton `desilog.log.log`.

**Locally get a standard format logger at a non-default loglevel**
`get_logger(loglevel)` should return a DESI standard formatted logger at the requested loglevel, but without changing the global loglevel of the `desilog.log.log` singleton, e.g. to be used in a function where you want to optionally turn on debugging messages without changing the loglevel of every other piece of code.  `get_logger()` should return that singleton by default.
```python
def blat(x, y, debug=False):
    from desiutil.log import get_logger
    if debug:
        log = get_logger('DEBUG')
    else:
        log = get_logger()  #- returns the default desilog.log.log singleton
```

**Global loglevel override**
Allow scripts to override the global loglevel in a way that impacts all subsequent usage of `desiutil.log.log`, even if that had already been previously imported:
```python
def main():
    # [parse arguments...]
    if args.debug:
        desiutil.log.set_loglevel('DEBUG')
```
i.e. this would need to change the behavior of the `desiutil.log.log` singleton itself, not replace the singleton with a different object so that if a previous import had already grabbed a reference to that singleton, its behavior would also change.  I think this is possible, but if we get bogged down in implementation details we could abandon the simplicity of the singleton and always require a fresh call to `get_logger()`.

**Other options like timestamps**
Local in a function:
```python
log = get_logger(timestamps=True)
```
Or globally, e.g. to be used by a script `main()`
```python
desiutil.log.use_timestamps(args.timestamps)  #- impacts all future logger calls
...
log = get_logger()               #- would include timestamps
desiutil.log.log.error('blarg')  #- would also include timestamps
```

**Comment by weaverba137** [comment]
Thanx for the comments.  None of this looks especially hard, though I think this might have to be lower priority due to other items going on right now.

**Comment by weaverba137** [comment]
Looking back over the last round of suggestions, the suggested *default* behavior would be:
```python
from desiutil.log import log
log.info('blat foo')
log.debug('biz bat')
```
That's fine, but most DESI code won't actually use this default unless we actually go back and change it.  The suggested secondary behavior would be equivalent to:
```python
from desiutil.log import get_logger
log = get_logger()
```
But that's indistinguishable from the existing API.

So we could implement this change, but nobody would use it, because everyone is already using the secondary/alternative API, and nothing would break.  It would be more work for the *users* of the module, in addition to the work of implementation.

**Comment by sbailey** [comment]
We're not yet done writing code and converting all print -> log, so the new API would still be useful.  This issue also has requests for non-default behavior for changing the loglevel local to a piece of code without it impacting other later uses.  This would require some refactoring of how the logging is organized, which might make the requested default behavior easy anyway (basically having a pre-created singleton with default logging parameters).

At the same time, let's not get stuck on the `from desiutil.log import log; log.info('blat foo')` case.  The more substantive request is to be able to get a DEBUG-level logger local to a piece of code without it impacting all later uses of logging.  i.e. local use of logging shouldn't change global logging state.

**Comment by weaverba137** [comment]
OK, that sounds like a good first step.

**Comment by weaverba137** [comment]
One more thing: do we want to tweak the behavior of `DESI_LOGLEVEL` as part of this?  Currently `DESI_LOGLEVEL` overrides *any* level argument.

**Comment by sbailey** [comment]
I think this is the order of overrides/priorities that we want:
```python
def get_logger(loglevel=None):
    if loglevel is None:
        if 'DESI_LOGLEVEL' in os.environ:
            loglevel = os.environ['DESI_LOGLEVEL']
        else:
            loglevel = 'INFO'
    # ...
```

i.e.
1. INFO is the default if $DESI_LOGLEVEL *isn't* set,
2. $DESI_LOGLEVEL overrides that default,
3. if user code calls `get_logger(loglevel)` with a non-None loglevel, use that.

Most user code would *not* use (3), except to temporarily turn on DEBUG logging within a piece of code while not using $DESI_LOGLEVEL that would turn it on for every piece of code.

**Comment by sbailey** [comment]
Reopening because PR #92 didn't fully address the requests here (but github automatically closed it based on parsing the PR comments...).  In particular, calling get_logger() with a non-default loglevel still has the side effect of changing the log level for all other log uses elsewhere in the code.

**Comment by weaverba137** [comment]
Just to clarify, is that the *only* outstanding problem?  Because if it is not, I would prefer to close this issue, and reopen new issues on the specific problems.

**Comment by sbailey** [comment]
Please take a look at the long May 11 comment with requests and use cases.  I think that none of those are addressed yet.  PR #92 did fix the relative priority of `$DESI_LOGLEVEL`, added more tests, and included the new feature of a context manager, but it didn't address the original request of this ticket that getting a non-default loglevel shouldn't have side effects.  You might argue that the other requests are low priority and could be spun off to other tickets, but let's keep this one open until the original request is fixed.

**Comment by weaverba137** [comment]
I have an idea to address the singleton request and the non-interference.  Question though: should the *singleton* be affected by `$DESI_LOGLEVEL`, or should that have an effectively immutable configuration?

**Comment by sbailey** [comment]
If we continue to have to use `get_logger()`, then I think it should pay attention to `$DESI_LOGLEVEL` at the time that `get_logger()` was called, but ignore any post-facto changes.  e.g.
```python
os.environ['DESI_LOGLEVEL'] = 'DEBUG'
log1 = get_logger()   #- returns a debug-level logger
log1.debug('blat')    #- prints blat
os.environ['DESI_LOGLEVEL'] = 'INFO'
log1.debug('foo')     #- prints foo even though $DESI_LOGLEVEL=INFO now
log2 = get_logger()   #- returns an info-level logger
log2.debug('quat')    #- silent
```
I'm pretty sure that this would work, possibly with minor modifications for string vs. integer loglevel and making sure that we don't add N>>1 identical handlers after N>>1 calls to `get_logger`:
```python
def get_logger(loglevel=None):
    if loglevel is None:
        if 'DESI_LOGLEVEL' in os.environ:
            loglevel = os.environ['DESI_LOGLEVEL']
        else:
            loglevel = 'INFO'
    log = logging.getLogger('DESI.{}'.format(loglevel))
    log.setLevel(loglevel)
    #- proceed with configuring formatting and handlers
    ...
    return log
```
i.e. `logging.getLogger` will be returning singletons, but a different singleton per loglevel.

**Comment by weaverba137** [comment]
@sbailey, you've hit on more-or-less exactly the method I'm using to deal with the non-interference problem.
